### ISSUE TITLE
gradcheck does not work for self-overlapping inputs 

### ISSUE THREAD

**Issue by kshitij12345** [pytorch/pytorch#38586]
The following code fails on CPU as well as GPU
```python
import torch
from torch.autograd import gradcheck

mat = torch.randn(2, 2, dtype=torch.float64)
vec = torch.randn(1, dtype=torch.float64).expand(2).requires_grad_(True)

def fn(vec):
    return (mat @ vec).sum()

gradcheck(fn, (vec,))

```

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by ngimel** [comment]
This is an error in gradcheck rather than in gradient computation. It's not specific to mv and would happen for self-overlapping inputs.
